File: verify/call/bsearch.py
def search(arr, item):
    # (NOT a bsearch)
    return arr.index(item)

def is_inc(arr):
    for i in range(1, len(arr)):
        if arr[i-1] >= arr[i]:
            return False
    return True

def expect(*args):
    i = 1
    n = len(args)
    items = []
    while i < n:
        if args[i] == "--key":
            key = int(args[i+1])
            i += 1
        else:
            items.append(int(args[i]))
        i += 1
    if not is_inc(items):
        return (1, "", "List is not increasing")
    try:
        p = search(items, key)
        return (0, "Found item = {} at position {}\n".format(key, p), "")
    except ValueError:
        return (0, "Item = {} could not be found\n".format(key), "")

data = [
    "1 2 3 5 8 9 --key 10".split(),
    "1 2 5 --key 2 6 7".split(),
    "5 4 3".split(),
    "--key 10 1".split(),
    "2 2 4 6 8".split(),
    "--key 5 1 3 5 6 --key 7 8 9 --key 3".split(),
]

File: tests/call/bsearch.c
int cmp(int* a, int* b) {
   return (*a - *b);
}

void assert_increase(int* arr, int len) {
    for (int i = 1; i < len; i++) {
        assert(arr[i-1] < arr[i]);
    }
}

int main(int argc, char** argv) {
    int* items = malloc(QSIZE*argc);
    int key;
    int j = 0;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--key") == 0) {
            key = atol(argv[++i]);
        } else {
            items[j++] = atol(argv[i]);
        }
    }

    try {
        assert_increase(items, j);
    } catch (AssertionFailure) {
        fprintf(stderr, "List is not increasing");
        exit(1);
    }

    int* item = bsearch(&key, items, j, QSIZE, cmp);
    if (item != NULL) {
        printf("Found item = %d at position %d\n", *item, (item - items) / QSIZE);
    } else {
        printf("Item = %d could not be found\n", key);
    }

    free(items);
    return 0;
}
